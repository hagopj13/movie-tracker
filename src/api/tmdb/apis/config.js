// @flow
import axios from '../axiosInstance';

export const getConfig = () => axios.get('/configuration');

export const getAllGenres = () => axios.get('/genre/movie/list');
