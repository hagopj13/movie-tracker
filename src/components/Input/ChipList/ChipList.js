// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Chip from '@material-ui/core/Chip';

type Props = {
  items: Array<{ id: number, name: string }>,
  selectedItems: number[],
  onItemClick: (id: number) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    margin: theme.spacing(-1, 0, 0, -1),
    '& > *': {
      margin: theme.spacing(1, 0, 0, 1),
    },
  },
  chipLabel: {
    fontSize: 12,
  },
  selectedChip: {
    color: theme.palette.common.white,
    border: `1px solid ${theme.palette.primary.main}`,
  },
}));

const ChipList = (props: Props) => {
  const { items, selectedItems, onItemClick } = props;

  const classes = useStyles();

  const isItemSelected = (id: number) => selectedItems.includes(id);

  return (
    <div className={classes.root}>
      {items.map((item) => (
        <Chip
          key={item.id}
          className={isItemSelected(item.id) ? classes.selectedChip : null}
          classes={{ label: classes.chipLabel }}
          label={item.name}
          color={isItemSelected(item.id) ? 'primary' : 'default'}
          variant={isItemSelected(item.id) ? 'default' : 'outlined'}
          onClick={() => onItemClick(item.id)}
        />
      ))}
    </div>
  );
};

export default ChipList;
