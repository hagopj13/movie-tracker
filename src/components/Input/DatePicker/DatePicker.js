// @flow
import React from 'react';
import type { Moment } from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import ClearIcon from '@material-ui/icons/Clear';
import { KeyboardDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

type Props = {
  label: string,
  selectedDate: Moment,
  onDateChange: (date: Moment) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    marginLeft: theme.spacing(0.5),
  },
  label: {
    minWidth: theme.spacing(5),
    width: theme.spacing(5),
  },
  datePicker: {
    width: '100%',
  },
  icon: {
    padding: 0,
  },
}));

const DatePicker = (props: Props) => {
  const { label, selectedDate, onDateChange } = props;

  const classes = useStyles();

  const handleDateChange = (date: Moment) => {
    if (date !== selectedDate && !date?.isSame(selectedDate)) {
      onDateChange(date);
    }
  };

  return (
    <div className={classes.root}>
      <Typography className={classes.label}>{label}</Typography>
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <KeyboardDatePicker
          className={classes.datePicker}
          disableToolbar
          variant="inline"
          inputVariant="outlined"
          format="DD/MM/YYYY"
          margin="dense"
          color="primary"
          value={selectedDate}
          onChange={handleDateChange}
          InputProps={{
            endAdornment: (
              <IconButton className={classes.icon} onClick={() => handleDateChange(null)}>
                <ClearIcon />
              </IconButton>
            ),
          }}
          InputAdornmentProps={{
            position: 'start',
          }}
          KeyboardButtonProps={{
            className: classes.icon,
          }}
        />
      </MuiPickersUtilsProvider>
    </div>
  );
};

export default DatePicker;
