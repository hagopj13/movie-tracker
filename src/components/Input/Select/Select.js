// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import MuiSelect from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';

type Props = {
  options: Array<{ value: string, label: string }>,
  selectedValue: string,
  onValueChange: (value: string) => void,
};

const useStyles = makeStyles({
  root: {
    width: '100%',
  },
});

const Select = (props: Props) => {
  const { options, selectedValue, onValueChange } = props;

  const classes = useStyles();

  const handleValueChange = (event) => {
    onValueChange(event.target.value);
  };

  return (
    <MuiSelect
      className={classes.root}
      value={selectedValue}
      onChange={handleValueChange}
      variant="outlined"
      margin="dense"
    >
      {options.map((option) => (
        <MenuItem key={option.value} value={option.value}>
          {option.label}
        </MenuItem>
      ))}
    </MuiSelect>
  );
};

export default Select;
