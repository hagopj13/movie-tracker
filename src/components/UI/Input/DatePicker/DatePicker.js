// @flow
import React from 'react';
import type { Moment } from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import ClearIcon from '@material-ui/icons/Clear';
import { KeyboardDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

type Props = {
  label: string,
  selectedDate: Moment,
  onDateChange: (date: Moment) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    marginLeft: theme.spacing(0.5),
  },
  label: {
    minWidth: 40,
    width: 40,
    fontWeight: 300,
  },
  icon: {
    padding: 0,
  },
}));

const DatePicker = (props: Props) => {
  const { label, selectedDate, onDateChange } = props;

  const classes = useStyles();

  return (
    <div className={classes.root}>
      {label && <Typography className={classes.label}>{label}</Typography>}
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <KeyboardDatePicker
          disableToolbar
          variant="inline"
          inputVariant="outlined"
          format="DD/MM/YYYY"
          margin="dense"
          color="primary"
          value={selectedDate}
          onChange={onDateChange}
          InputProps={{
            endAdornment: (
              <IconButton className={classes.icon} onClick={() => onDateChange(null)}>
                <ClearIcon />
              </IconButton>
            ),
          }}
          InputAdornmentProps={{
            position: 'start',
          }}
          KeyboardButtonProps={{
            className: classes.icon,
          }}
        />
      </MuiPickersUtilsProvider>
    </div>
  );
};

export default DatePicker;
