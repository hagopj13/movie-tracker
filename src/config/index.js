import React from 'react';
import DiscoverIcon from '@material-ui/icons/Movie';
import UpcomingIcon from '@material-ui/icons/Update';
import ProfileIcon from '@material-ui/icons/Person';

export const navigationItems = [
  {
    id: 'DISCOVER',
    label: 'Discover',
    to: '/',
    icon: <DiscoverIcon />,
    requiresAuth: false,
  },
  {
    id: 'UPCOMING',
    label: 'Upcoming',
    to: '/upcoming',
    icon: <UpcomingIcon />,
    requiresAuth: false,
  },
  {
    id: 'PROFILE',
    label: 'Profile',
    to: '/profile',
    icon: <ProfileIcon />,
    requiresAuth: true,
  },
];

export const dialogTypes = {
  LOGIN: 'LOGIN',
};

export const sortOptions = [
  { value: 'popularity.desc', label: 'Popularity Descending' },
  { value: 'popularity.asc', label: 'Popularity Ascending' },
  { value: 'vote_average.desc', label: 'Rating Descending' },
  { value: 'vote_average.asc', label: 'Rating Ascending' },
  { value: 'release_date.desc', label: 'Release Date Descending' },
  { value: 'release_date.asc', label: 'Release Date Ascending' },
];
