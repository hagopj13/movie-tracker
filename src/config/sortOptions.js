export default [
  { value: 'popularity.desc', label: 'Popularity Descending' },
  { value: 'popularity.asc', label: 'Popularity Ascending' },
  { value: 'vote_average.desc', label: 'Rating Descending' },
  { value: 'vote_average.asc', label: 'Rating Ascending' },
  { value: 'release_date.desc', label: 'Release Date Descending' },
  { value: 'release_date.asc', label: 'Release Date Ascending' },
];
