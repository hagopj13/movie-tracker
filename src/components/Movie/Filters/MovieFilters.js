// @flow
import React from 'react';
import type { Moment } from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';

import FilterBox from 'components/Filter/Box/FilterBox';
import FilterBoxItem from 'components/Filter/Box/Item/FilterBoxItem';
import Select from 'components/Input/Select/Select';
import ChipList from 'components/Input/ChipList/ChipList';
import DatePicker from 'components/Input/DatePicker/DatePicker';
import sortOptions from 'config/sortOptions';
import type { State as FiltersState } from 'store/common/filters/filters.reducer';
import type { Genre } from 'types';

type Props = {
  allGenres: Genre[],
  selectedFilters: FiltersState,
  onFiltersChanged: () => void,
  onSetSortBy: (sortBy: string) => void,
  onToggleGenre: (genreId: number) => void,
  onSetReleaseDateStart: (releaseDateStart: Moment) => void,
  onSetReleaseDateEnd: (releaseDateEnd: Moment) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    '& > *': {
      marginTop: theme.spacing(3),
    },
    '& > :first-child': {
      marginTop: 0,
    },
  },
}));

const MovieFilters = (props: Props) => {
  const {
    allGenres,
    selectedFilters,
    onFiltersChanged,
    onSetSortBy,
    onToggleGenre,
    onSetReleaseDateStart,
    onSetReleaseDateEnd,
  } = props;

  const classes = useStyles();

  const handleSetSortBy = (sortBy) => {
    onSetSortBy(sortBy);
    onFiltersChanged();
  };

  const handleToggleGenre = (genreId) => {
    onToggleGenre(genreId);
    onFiltersChanged();
  };

  const handleSetReleaseDateStart = (releaseDateStart) => {
    onSetReleaseDateStart(releaseDateStart);
    onFiltersChanged();
  };

  const handleSetReleaseDateEnd = (releaseDateEnd) => {
    onSetReleaseDateEnd(releaseDateEnd);
    onFiltersChanged();
  };

  return (
    <div className={classes.root}>
      <FilterBox title="Sort">
        <FilterBoxItem title="Sort by">
          <Select
            options={sortOptions}
            selectedValue={selectedFilters.sortBy}
            onValueChange={handleSetSortBy}
          />
        </FilterBoxItem>
      </FilterBox>
      <FilterBox title="Filters">
        <FilterBoxItem title="Release date">
          <DatePicker
            label="Start"
            selectedDate={selectedFilters.releaseDateStart}
            onDateChange={handleSetReleaseDateStart}
          />
          <DatePicker
            label="End"
            selectedDate={selectedFilters.releaseDateEnd}
            onDateChange={handleSetReleaseDateEnd}
          />
        </FilterBoxItem>
        <Divider light />
        <FilterBoxItem title="Genres">
          <ChipList
            items={allGenres}
            selectedItems={selectedFilters.genres}
            onItemClick={handleToggleGenre}
          />
        </FilterBoxItem>
      </FilterBox>
    </div>
  );
};

export default MovieFilters;
