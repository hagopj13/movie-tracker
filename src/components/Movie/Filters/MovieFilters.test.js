import React from 'react';
import moment from 'moment';
import { createShallow } from '@material-ui/core/test-utils';

import { genres } from 'store/fixtures/config';
import sortOptions from 'config/sortOptions';
import Select from 'components/Input/Select/Select';
import ChipList from 'components/Input/ChipList/ChipList';
import DatePicker from 'components/Input/DatePicker/DatePicker';

import MovieFilters from './MovieFilters';

describe('MovieFilters component', () => {
  let shallow;
  let allGenres;
  let selectedFilters;
  let wrapper;

  const mockFiltersChanged = jest.fn();
  const mockSetSortBy = jest.fn();
  const mockToggleGenre = jest.fn();
  const mockSetReleaseDateStart = jest.fn();
  const mockSetReleaseDateEnd = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    allGenres = genres;
    selectedFilters = {
      sortBy: sortOptions[1].value,
      genres: [genres[1].id],
      releaseDateStart: moment('2020-01-01'),
      releaseDateEnd: moment('2020-02-01'),
    };
    const mockProps = {
      allGenres,
      selectedFilters,
      onFiltersChanged: mockFiltersChanged,
      onSetSortBy: mockSetSortBy,
      onToggleGenre: mockToggleGenre,
      onSetReleaseDateStart: mockSetReleaseDateStart,
      onSetReleaseDateEnd: mockSetReleaseDateEnd,
    };
    wrapper = shallow(<MovieFilters {...mockProps} />);
  });

  it('should render the MovieFilters component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass sortBy selected filter to Select', () => {
    expect(wrapper.find(Select).prop('selectedValue')).toBe(selectedFilters.sortBy);
  });

  it('should pass genres selected filter to ChipList', () => {
    expect(wrapper.find(ChipList).prop('selectedItems')).toBe(selectedFilters.genres);
  });

  it('should pass releaseDateStart selected filter to DatePicker', () => {
    expect(wrapper.find(DatePicker).at(0).prop('selectedDate')).toBe(
      selectedFilters.releaseDateStart,
    );
  });

  it('should pass releaseDateEnd selected filter to DatePicker', () => {
    expect(wrapper.find(DatePicker).at(1).prop('selectedDate')).toBe(
      selectedFilters.releaseDateEnd,
    );
  });

  it('should handle sortBy filter change', () => {
    const newSortBy = sortOptions[2].value;
    wrapper.find(Select).prop('onValueChange')(newSortBy);
    expect(mockSetSortBy).toHaveBeenCalledTimes(1);
    expect(mockSetSortBy).toHaveBeenLastCalledWith(newSortBy);
    expect(mockFiltersChanged).toHaveBeenCalledTimes(1);
  });

  it('should handle genre toggle', () => {
    const genre = genres[2].id;
    wrapper.find(ChipList).prop('onItemClick')(genre);
    expect(mockToggleGenre).toHaveBeenCalledTimes(1);
    expect(mockToggleGenre).toHaveBeenLastCalledWith(genre);
    expect(mockFiltersChanged).toHaveBeenCalledTimes(1);
  });

  it('should handle releaseDateStart filter change', () => {
    const newReleaseDateStart = moment();
    wrapper.find(DatePicker).at(0).prop('onDateChange')(newReleaseDateStart);
    expect(mockSetReleaseDateStart).toHaveBeenCalledTimes(1);
    expect(mockSetReleaseDateStart).toHaveBeenLastCalledWith(newReleaseDateStart);
    expect(mockFiltersChanged).toHaveBeenCalledTimes(1);
  });

  it('should handle releaseDateEnd filter change', () => {
    const newReleaseDateEnd = moment();
    wrapper.find(DatePicker).at(1).prop('onDateChange')(newReleaseDateEnd);
    expect(mockSetReleaseDateEnd).toHaveBeenCalledTimes(1);
    expect(mockSetReleaseDateEnd).toHaveBeenLastCalledWith(newReleaseDateEnd);
    expect(mockFiltersChanged).toHaveBeenCalledTimes(1);
  });
});
