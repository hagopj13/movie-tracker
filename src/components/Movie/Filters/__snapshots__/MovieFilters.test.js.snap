// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieFilters component should render the MovieFilters component correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <FilterBox
    title="Sort"
  >
    <FilterBox
      title="Sort by"
    >
      <Select
        onValueChange={[Function]}
        options={
          Array [
            Object {
              "label": "Popularity Descending",
              "value": "popularity.desc",
            },
            Object {
              "label": "Popularity Ascending",
              "value": "popularity.asc",
            },
            Object {
              "label": "Rating Descending",
              "value": "vote_average.desc",
            },
            Object {
              "label": "Rating Ascending",
              "value": "vote_average.asc",
            },
            Object {
              "label": "Release Date Descending",
              "value": "release_date.desc",
            },
            Object {
              "label": "Release Date Ascending",
              "value": "release_date.asc",
            },
          ]
        }
        selectedValue="popularity.asc"
      />
    </FilterBox>
  </FilterBox>
  <FilterBox
    title="Filters"
  >
    <FilterBox
      title="Release date"
    >
      <DatePicker
        label="Start"
        onDateChange={[Function]}
        selectedDate={"2019-12-31T23:00:00.000Z"}
      />
      <DatePicker
        label="End"
        onDateChange={[Function]}
        selectedDate={"2020-01-31T23:00:00.000Z"}
      />
    </FilterBox>
    <WithStyles(ForwardRef(Divider))
      light={true}
    />
    <FilterBox
      title="Genres"
    >
      <ChipList
        items={
          Array [
            Object {
              "id": 28,
              "name": "Action",
            },
            Object {
              "id": 12,
              "name": "Adventure",
            },
            Object {
              "id": 16,
              "name": "Animation",
            },
            Object {
              "id": 35,
              "name": "Comedy",
            },
          ]
        }
        onItemClick={[Function]}
        selectedItems={
          Array [
            12,
          ]
        }
      />
    </FilterBox>
  </FilterBox>
</div>
`;
